perf_hooks: performance milestone time origin timestamp improvement

IlyasShabi opened the pull request:
local benchmark:

```                              
                                                        confidence improvement accuracy (*)   (**)  (***)
perf_hooks/time-origin.js method='timeOrigin' n=1000000        ***     75.84 %       ±1.07% ±1.44% ±1.91%
perf_hooks/time-origin.js method='toJSON' n=1000000            ***     27.41 %       ±1.24% ±1.66% ±2.16%

```

H4ad reviewed (approved):
Is always nice to see more usage of Fast APIs, thanks for the PR!

joyeecheung commented:
Not blocking but just a reminder - it would be faster if we just pass `env->time_origin_timestamp()` the same we we pass `env->time_origin()` to JS land via a typed array, and don't do a native call at all, because this is conceptually a constant (I am a bit surprised that it's not done already since we've been doing this for `env->time_origin()`, though with a closer look it seems to be only used by `performance.now()`.

IlyasShabi commented:
Thanks for the suggestion @joyeecheung ! I've changed the code to pass `env->time_origin_timestamp()` the same way we do with `env->time_origin()`. It's now faster. Here is the benchmarks to show the improvement.

```
                                                        confidence improvement accuracy (*)   (**)  (***)
perf_hooks/time-origin.js method='timeOrigin' n=1000000        ***    120.57 %       ±2.46% ±3.30% ±4.35%
perf_hooks/time-origin.js method='toJSON' n=1000000            ***     32.10 %       ±1.64% ±2.18% ±2.86%
```

I've updated the commit message and PR title since it's no more a Fast api implementation, could you please review it again 🙏

joyeecheung commented on src/node_perf.cc:
```diff
@@ -43,6 +42,7 @@ uint64_t performance_v8_start;
 
 PerformanceState::PerformanceState(Isolate* isolate,
                                    uint64_t time_origin,
+                                   uint64_t time_origin_timestamp,
```
```suggestion
                                   double time_origin_timestamp,
```

joyeecheung commented on src/node_perf.cc:
```diff
@@ -86,23 +86,27 @@ PerformanceState::SerializeInfo PerformanceState::Serialize(
-  // not use the Mark() method. The time origin milestone is not exposed
-  // to user land.
+void PerformanceState::Initialize(uint64_t time_origin,
+                                  uint64_t time_origin_timestamp) {
```
```suggestion
                                  double time_origin_timestamp) {
```

joyeecheung commented on src/node_perf.cc:
```diff
@@ -86,23 +86,27 @@ PerformanceState::SerializeInfo PerformanceState::Serialize(
   this->milestones[NODE_PERFORMANCE_MILESTONE_TIME_ORIGIN] =
       static_cast<double>(time_origin);
+  this->milestones[NODE_PERFORMANCE_MILESTONE_TIME_ORIGIN_TIMESTAMP] =
+      static_cast<double>(time_origin_timestamp);
```
```suggestion
     time_origin_timestamp;
```

joyeecheung commented on src/node_perf.cc:
```diff
@@ -86,23 +86,27 @@ PerformanceState::SerializeInfo PerformanceState::Serialize(
 void PerformanceState::Deserialize(v8::Local<v8::Context> context,
-                                   uint64_t time_origin) {
+                                   uint64_t time_origin,
+                                   uint64_t time_origin_timestamp) {
```
```suggestion
                                   double time_origin_timestamp) {
```

joyeecheung commented on src/node_perf_common.h:
```diff
@@ -64,10 +65,13 @@ class PerformanceState {
-  void Deserialize(v8::Local<v8::Context> context, uint64_t time_origin);
+  void Deserialize(v8::Local<v8::Context> context,
+                   uint64_t time_origin,
+                   uint64_t time_origin_timestamp);
```
```suggestion
                   double time_origin_timestamp);
```

joyeecheung commented on src/node_perf_common.h:
```diff
@@ -81,7 +85,7 @@ class PerformanceState {
 
  private:
-  void Initialize(uint64_t time_origin);
+  void Initialize(uint64_t time_origin, uint64_t time_origin_timestamp);
```
```suggestion
  void Initialize(uint64_t time_origin, double time_origin_timestamp);
```

joyeecheung reviewed (approved):
LGTM % comments. The timestamp is already a double and there is no need to cast it to uint64_t on the way.

IlyasShabi commented:
Failed on this one `wasi/test-wasi # TODO : Fix flaky test`
@joyeecheung should I do something to merge this one ?

IlyasShabi commented:
Could someone merge this ? 👀
